Source/NSDictionaryController.m

```diff
@@ -0,0 +1,480 @@
+   Date: 16-10-2021
+
+   This file is part of the GNUstep Library.
+
+   This library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   This library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; if not, write to the Free
+   Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+   Boston, MA 02110 USA.
+*/
+
+#import <Foundation/NSArray.h>
+#import <Foundation/NSDictionary.h>
+#import <Foundation/NSException.h>
+#import <Foundation/NSIndexSet.h>
+#import <Foundation/NSKeyValueObserving.h>
+#import <Foundation/NSPropertyList.h>
+#import <Foundation/NSString.h>
+
+#import "AppKit/NSDictionaryController.h"
+#import "AppKit/NSKeyValueBinding.h"
+
+#import "GSBindingHelpers.h"
+#import "GSFastEnumeration.h"
+
+// Private methods for NSString and NSDictionary...
+@interface NSDictionary (GSPrivate_DictionaryToStrings)
+
+- (NSString *) _stringsFromDictionary;
+
+@end
+
+@implementation NSDictionary (GSPrivate_DictionaryToStrings)
+
+- (NSString *) _stringsFromDictionary
+{
+  NSEnumerator *en = [self keyEnumerator];
+  NSString *k = nil;
+  NSString *result = @"";
+
+  while ((k = [en nextObject]) != nil)
+    {
+      NSString *v = [self objectForKey: k];
+      result = [result stringByAppendingString:
+			 [NSString stringWithFormat: @"\"%@\" = \"%@\";\n", k, v]];
+    }
+
+  return result;
+}
+
+@end
+
+@interface NSString (GSPrivate_StringsToDictionary)
+
+- (NSDictionary *) _dictionaryFromStrings;
+
+@end
+
+@implementation NSString (GSPrivate_StringsToDictionary)
+
+- (NSDictionary *) _dictionaryFromStrings
+{
+  NSError *error = nil;
+  NSDictionary *dictionary =
+    [NSPropertyListSerialization
+      propertyListWithData: [self dataUsingEncoding: NSUTF8StringEncoding]
+		   options: NSPropertyListImmutable
+		    format: NULL
+		     error: &error];
+  if (error != nil)
+    {
+      NSLog(@"Error reading strings file: %@", error);
+    }
+
+  return dictionary;
+}
+
+@end
+
+@interface NSDictionaryController (GSPrivate_BuildArray)
+
+- (NSArray *) _buildArray: (NSDictionary *)content;
+
+@end
+
+@implementation NSDictionaryController (GSPrivate_BuildArray)
+
+- (NSArray *) _buildArray: (NSDictionary *)content
+{
+  NSArray *allKeys = [content keysSortedByValueUsingSelector: @selector(compare:)];
+  NSMutableArray *result = [NSMutableArray arrayWithCapacity: [allKeys count]];
+
+  FOR_IN(id, k, allKeys)
+    {
+      NSDictionaryControllerKeyValuePair *kvp =
+	AUTORELEASE([[NSDictionaryControllerKeyValuePair alloc] init]);
```

Comment:
You could avoid the AUTORELEASE here by using a RELEASE after line 128 and in the skip case

Reply:
I feel like the AUTORELEASE addresses both cases, so it should be fine.   I think it keeps the code a little cleaner.  Let me know if you really would like me to do this the other way and I will do it as a quick fix.